dwarf-rust-experimental/rust/src/main.rs

```diff
@@ -0,0 +1,502 @@
+    |_| Ok(Default::default()),
+  )?;
+
+  let mut scored_source_units = vec![];
+
+  let mut iter = dwarf.units();
+  let mut best_scores: HashMap<u64, Line> = HashMap::new();
+  let mut next_entry_id = 0u64;
+  let mut next_file_id = 0u32;
+  let mut functions: BTreeMap<String, Vec<u64>> = BTreeMap::new();
+
+  while let Some(unit) = iter.next()? {
+    let mut unit = dwarf.unit(unit)?;
+
+    let line_program = match unit.line_program.take() {
+      Some(line_program) => line_program,
+      None => continue,
+    };
+
+    let lang = {
+      let mut entries = unit.entries();
+      entries.next_entry()?;
+      match entries
+        .current()
+        .unwrap()
+        .attr_value(gimli::DW_AT_language)?
+      {
+        Some(gimli::AttributeValue::Language(lang)) => lang.0,
+        _ => 0,
+      }
+    };
+
+    // Extract function information from DWARF entries
+    {
+      let mut entries = unit.entries();
+      while let Some((_depth, entry)) = entries.next_dfs()? {
+        if entry.tag() == gimli::DW_TAG_subprogram {
+          // Get function name
+          let func_name = if let Some(attr) = entry.attr_value(gimli::DW_AT_name)? {
+            match attr {
+              gimli::AttributeValue::DebugStrRef(offset) => {
+                match dwarf.debug_str.get_str(offset) {
+                  Ok(name_str) => Some(name_str.to_string()?.to_string()),
+                  Err(_) => None,
+                }
+              }
+              gimli::AttributeValue::String(name_str) => {
+                Some(name_str.to_string()?.to_string())
+              }
+              _ => None,
+            }
+          } else {
+            None
+          };
+          
+
+          // Get low_pc (start address) and high_pc (end address or size)
+          let low_pc = if let Some(attr) = entry.attr_value(gimli::DW_AT_low_pc)? {
+            match attr {
+              gimli::AttributeValue::Addr(addr) => Some(addr),
+              _ => None,
+            }
+          } else {
+            None
+          };
+
+          let high_pc = if let Some(attr) = entry.attr_value(gimli::DW_AT_high_pc)? {
+            match attr {
+              gimli::AttributeValue::Addr(addr) => Some(addr),
+              gimli::AttributeValue::Udata(offset) => {
+                // high_pc can be an offset from low_pc
+                if let Some(low) = low_pc {
+                  Some(low + offset)
+                } else {
+                  None
+                }
+              }
+              _ => None,
+            }
+          } else {
+            None
+          };
+
+          // If we have all the information, add it to our functions map
+          if let (Some(func_name), Some(start), Some(end)) = (func_name, low_pc, high_pc) {
+            // Filter out synthetic/relocated addresses similar to line processing
+            const MAX_REALISTIC_WASM_ADDR: u64 = 0x40000000; // 1GB threshold
```

Comment:
Why is this needed? Have you found related issues?

Reply:
Do you happen to know where are those coming from? Synthetic functions?
Should we just drop them when loading?